Source/WebCore/page/Navigation.cpp

```diff
@@ -171,34 +180,65 @@ NavigationAPIMethodTracker Navigation::maybeSetUpcomingNonTraversalTracker(Ref<D
+    apiMethodTracker.finishedPromise->reject(RejectAsHandled::Yes);
+
+    m_upcomingTraverseMethodTrackers.add(key, apiMethodTracker);
+
+    return apiMethodTracker;
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker-derived-result
+Navigation::Result Navigation::apiMethodTrackerDerivedResult(NavigationAPIMethodTracker& apiMethodTracker)
+{
+    ASSERT(apiMethodTracker.committedPromise);
+    Ref committed = *apiMethodTracker.committedPromise.leakRef();
+    Ref finished = *apiMethodTracker.finishedPromise.leakRef();
+    return {
+        createDOMPromise(committed),
+        createDOMPromise(finished),
+    };
+}
+
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload
 Navigation::Result Navigation::reload(ReloadOptions&& options, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
 {
+    if (!frame())
+        return createErrorResult(WTFMove(committed), WTFMove(finished), ExceptionCode::InvalidStateError, "Invalid state"_s);
+
     auto serializedState = serializeState(options.state);
     if (serializedState.hasException())
-        return createErrorResult(committed, finished, serializedState.releaseException());
+        return createErrorResult(WTFMove(committed), WTFMove(finished), serializedState.releaseException());
+
+    // Frame was detached during the serialize step.
+    if (!frame())
+        return createErrorResult(WTFMove(committed), WTFMove(finished), ExceptionCode::InvalidStateError, "Invalid state"_s);
 
     auto apiMethodTracker = maybeSetUpcomingNonTraversalTracker(WTFMove(committed), WTFMove(finished), WTFMove(options.info), serializedState.releaseReturnValue());
 
     // FIXME: Only a stub to reload for testing.
-    window()->frame()->loader().reload();
+    frame()->loader().reload();
 
-    // FIXME: keep track of promises to resolve later.
-    Ref entry = NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), { });
-    Navigation::Result result = { apiMethodTracker.committedPromise.ptr(), apiMethodTracker.finishedPromise.ptr() };
-    committed->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
-    finished->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
-    return result;
+    return apiMethodTrackerDerivedResult(apiMethodTracker);
 }
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate
-Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& options, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
+Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& options, Ref<DeferredPromise>&&committed, Ref<DeferredPromise>&&finished)
 {
-    auto currentURL = scriptExecutionContext()->url();
-    auto newURL = URL { currentURL, url };
+
+    if (!frame())
+        return createErrorResult(WTFMove(committed), WTFMove(finished), ExceptionCode::InvalidStateError, "Invalid state"_s);
```

Comment:
It is probably if I move these kind of detach fixes to https://github.com/WebKit/WebKit/pull/25013. WDYT?

Reply:
Yeah that would be fine.